### PR TITLE
hatch: 1.9.0 -> 1.9.7

### LINKED ISSUE #287216: Hatch check phase fails on darwin
### Describe the bug

Hatch doesn't build anymore. Seems a different issue than a year ago https://github.com/NixOS/nixpkgs/issues/209358

### Steps To Reproduce

This started failing again in `1.9.0`.

```
% nix shell nixpkgs#hatch                    
warning: Ignoring setting 'auto-allocate-uids' because experimental feature 'auto-allocate-uids' is not enabled
warning: Ignoring setting 'impure-env' because experimental feature 'configurable-impure-env' is not enabled
error: builder for '/nix/store/4mgjxh6ri2my6x0wfr2pmwfyv7qhnwkh-hatch-1.9.0.drv' failed with exit code 1;
       last 10 log lines:
       > FAILED tests/cli/run/test_run.py::test_sync_dependencies - AssertionError: assert 'Checking dependencies\n' == 'Checking dep...ependen...
       > FAILED tests/python/test_core.py::TestGetInstalled::test_order - hatch.errors.PythonDistributionResolutionError: Could not find a default so...
       > FAILED tests/cli/env/test_create.py::test_features - AssertionError: assert 1 == 2
       > FAILED tests/cli/run/test_run.py::test_sync_project_dependencies - AssertionError: assert 'Checking dependencies\n' == 'Checking dep...ependen...
       > FAILED tests/cli/run/test_run.py::test_sync_project_features - AssertionError: assert 'Checking dependencies\n' == 'Checking dep...ependen...
       > FAILED tests/cli/run/test_run.py::test_dependency_hash_checking - AssertionError: assert 'Checking dependencies\n' == 'Checking dep...ependen...
       > FAILED tests/cli/env/test_create.py::test_sync_dynamic_dependencies - AssertionError: assert 3 == 4
       > FAILED tests/cli/env/test_create.py::test_sync_dependencies - AssertionError: assert 'Creating env...ependencies\n' == 'Creating env...ep...
       > ================= 17 failed, 1628 passed, 55 skipped in 51.96s =================
       > /nix/store/ymqnr6q4slam0ns76mjpcjry70fa8h3i-stdenv-darwin/setup: line 1591: pop_var_context: head of shell_variables not a function context
       For full logs, run 'nix log /nix/store/4mgjxh6ri2my6x0wfr2pmwfyv7qhnwkh-hatch-1.9.0.drv'.
       ```
```

### Expected behavior
It builds

### Screenshots
If applicable, add screenshots to help explain your problem.

### Additional context

Related to https://github.com/NixOS/nixpkgs/pull/271586 ?

### Notify maintainers

@onny 


---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### LINKED ISSUE #308121: Build failure: hatch
### Steps To Reproduce

1. `nix shell nixpkgs#hatch`

### Build log

```
error: builder for '/nix/store/ka07hfapranqwyn75jmnxvw7vn6l3p49-hatch-1.9.0.drv' failed with exit code 1;
       last 25 log lines:
       > E       assert {} is {}
       > E        +  where {} = <tests.backend.builders.utils.MockBuilder object at 0x7ffff516af10>.project_config
       >
       > /build/hatch-1.9.0/tests/backend/builders/plugin/test_interface.py:65: AssertionError
       > ___________________________ TestDynamic.test_correct ___________________________
       > [gw15] linux -- Python 3.11.9 /nix/store/glfr70gi7hfaj50mwj2431p8bg60fhqw-python3-3.11.9/bin/python3.11
       >
       > self = <tests.backend.metadata.test_core.TestDynamic object at 0x7ffff4ec4710>
       > isolation = Path('/build/tmpnis8_81d')
       >
       >     def test_correct(self, isolation):
       >         dynamic = ['version']
       >         metadata = ProjectMetadata(str(isolation), None, {'project': {'dynamic': dynamic}})
       >     
       > >       assert metadata.core.dynamic is dynamic
       > E       AssertionError: assert ['version'] is ['version']
       > E        +  where ['version'] = <hatchling.metadata.core.CoreMetadata object at 0x7ffff55602d0>.dynamic
       > E        +    where <hatchling.metadata.core.CoreMetadata object at 0x7ffff55602d0> = <hatchling.metadata.core.ProjectMetadata object at 0x7ffff5740850>.core
       >
       > /build/hatch-1.9.0/tests/backend/metadata/test_core.py:87: AssertionError
       > =========================== short test summary info ============================
       > FAILED tests/backend/builders/plugin/test_interface.py::TestMetadata::test_core - assert {} is {}
       > FAILED tests/backend/metadata/test_core.py::TestDynamic::test_correct - AssertionError: assert ['version'] is ['version']
       > ================== 2 failed, 1603 passed, 99 skipped in 7.67s ==================
       > /nix/store/9wnvhjyxjykwn5y06xc9a2h8rs5fbfia-stdenv-linux/setup: line 1579: pop_var_context: head of shell_variables not a function context
       For full logs, run 'nix log /nix/store/ka07hfapranqwyn75jmnxvw7vn6l3p49-hatch-1.9.0.drv'.
```
### Additional context

I run into this problem recently after updating to latest unstable packages.

### Notify maintainers

@stuebinm @onny @mweinelt 

### Metadata

```
❯ nix-shell -p nix-info --run "nix-info -m"
this path will be fetched (0.00 MiB download, 0.00 MiB unpacked):
  /nix/store/mh3dkpww5py0fnrihk45pw7hvkj5fxxc-nix-info
copying path '/nix/store/mh3dkpww5py0fnrihk45pw7hvkj5fxxc-nix-info' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"`
 - host os: `Linux 6.8.4-200.fc39.x86_64, Fedora Linux, 39 (Container Image), nobuild`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.21.1`
 - nixpkgs: `/var/home/user/.nix-defexpr/channels/nixpkgs`
```

## Links
I think this might be related to https://github.com/NixOS/nixpkgs/pull/297084

### PR DESCRIPTION
**Pull request by onny**
## Description of changes

Superseeds: https://github.com/NixOS/nixpkgs/pull/283065
Fixes: https://github.com/NixOS/nixpkgs/issues/287216
Fixes: https://github.com/NixOS/nixpkgs/issues/308121

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by nim65s** [comment]
Result of `nixpkgs-review pr 308354` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>hatch</li>
    <li>hatch.dist</li>
  </ul>
</details>

**Review by nim65s — APPROVED** [review]

**Comment by getchoo** [comment]
Result of `nixpkgs-review pr 308354` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>hatch</li>
    <li>hatch.dist</li>
  </ul>
</details>

File: pkgs/by-name/ha/hatch/package.nix [hunk 1]
```diff
@@ -69,19 +74,13 @@ python3.pkgs.buildPythonApplication rec {
...
     "test_already_installed_update_flag"
     "test_already_installed_update_prompt"
-    # unmet expectations about the binary module we provide
-    "test_dependency_not_found"
-    "test_marker_unmet"
-    # output capturing mismatch, likely stdout/stderr mixup
-    "test_no_compatibility_check_if_exists"
+    # Loosen hatchling runtime version dependency
+    "test_core"
+    "test_correct"
```
Comment (getchoo) [inline pkgs/by-name/ha/hatch/package.nix:83]:
```suggestion
    "test_correct"
    "test_resolve"
```

not exactly sure why this fails on aarch64-linux but not x86_64

**Review by getchoo — CHANGES_REQUESTED** [review]

**Comment by onny** [comment]
Thank you @getchoo added the failing test to the list

**Pull request merged**